In [1]:
import warnings

import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

import missingno as msno

import statsmodels.api as sm
import statsmodels.formula.api as smf

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.metrics import roc_curve, roc_auc_score

import factor_analyzer as fact

import funciones_auxiliares_hito_3 as aux3
from funciones_auxiliares_hito_2 import preprocesar_y_recodificar_enunciado_dos

# warnings.filterwarnings('ignore')
sns.set_theme()
plt.rcParams['figure.figsize'] = (12, 6)

In [2]:
df = pd.read_csv('students.csv', delimiter='|')
df = preprocesar_y_recodificar_enunciado_dos(df)

# Analisis de factores latentes en variables famrel:health

In [3]:
posibles_factores_latentes = df.loc[:, 'famrel':'health']
print(fact.calculate_bartlett_sphericity(posibles_factores_latentes))

(264.4978629665067, 6.03816710126852e-48)


Por lo tanto, existe al menos una correlacion entre dos variables que es distinto de 0

In [4]:
fact.calculate_kmo(posibles_factores_latentes)

(array([0.50173343, 0.57692329, 0.61683967, 0.58664172, 0.5554092 ,
        0.59265861]),
 0.5767240731164592)

Como el valor del test de Kaiser Meyer Olkins es menor a 0.7, entonces NO existe un factor latente
en nuestros datos.

# Modelo para G1

In [7]:
modelo_g1 = aux3.crear_modelo_lineal(df.drop(columns=['G2', 'G3']), 'G1')
modelo_g1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     G1   R-squared:                       0.331
Model:                            OLS   Adj. R-squared:                  0.224
Method:                 Least Squares   F-statistic:                     3.090
Date:                Sat, 04 Feb 2023   Prob (F-statistic):           5.75e-08
Time:                        23:08:48   Log-Likelihood:                -683.63
No. Observations:                 284   AIC:                             1447.
Df Residuals:                     244   BIC:                             1593.
Df Model:                          39                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept          9.6531      3.452      2.797      0.006       2.855      16.452
school            -0.3886      0.666     -0.583      0.560      -1.701       0.924
sex                0.7631      0.432      1.768      0.078      -0.087       1.613
age                0.0772      0.185      0.418      0.677      -0.287       0.441
address           -0.0205      0.487     -0.042      0.966      -0.980       0.939
famsize            0.3125      0.440      0.710      0.478      -0.555       1.180
Pstatus            0.1872      0.650      0.288      0.774      -1.094       1.468
Medu               0.1809      0.282      0.642      0.521      -0.374       0.736
Fedu               0.1527      0.236      0.646      0.519      -0.313       0.618
traveltime        -0.0215      0.291     -0.074      0.941      -0.594       0.551
studytime          0.5074      0.245      2.068      0.040       0.024       0.991
failures          -1.1235      0.282     -3.977      0.000      -1.680      -0.567
schoolsup         -1.7737      0.564     -3.145      0.002      -2.885      -0.663
famsup             1.0062      0.411      2.447      0.015       0.196       1.816
paid               0.0321      0.403      0.080      0.937      -0.762       0.826
activities         0.4719      0.370      1.276      0.203      -0.257       1.201
nursery            0.0255      0.455      0.056      0.955      -0.871       0.922
higher            -1.1077      0.876     -1.265      0.207      -2.833       0.617
internet          -0.0671      0.502     -0.134      0.894      -1.055       0.921
romantic          -0.0842      0.399     -0.211      0.833      -0.870       0.702
famrel             0.1449      0.213      0.681      0.497      -0.274       0.564
freetime           0.3778      0.193      1.959      0.051      -0.002       0.758
goout             -0.7691      0.199     -3.873      0.000      -1.160      -0.378
Dalc              -0.0730      0.286     -0.255      0.799      -0.636       0.490
Walc               0.1585      0.210      0.756      0.451      -0.255       0.572
health            -0.2918      0.142     -2.062      0.040      -0.571      -0.013
absences          -0.0048      0.024     -0.196      0.845      -0.053       0.043
health_Mjob        1.0251      0.918      1.116      0.265      -0.784       2.834
other             -0.5423      0.598     -0.907      0.365      -1.720       0.636
services           0.3913      0.676      0.579      0.563      -0.939       1.722
teacher           -1.0850      0.836     -1.298      0.195      -2.731       0.561
health_Fjob       -1.0572      1.304     -0.810      0.418      -3.626       1.512
other_Fjob        -1.4765      0.984     -1.500      0.135      -3.415       0.462
services_Fjob     -0.8524      0.990     -0.861      0.390      -2.801       1.097
teacher_Fjob

Al realizar un modelo de regresion lineal con todas las variables y G1 como vector objetivo se
obtuvo lo siguiente:

- El modelo presenta un $R^2$ = 0.331. Por lo tanto, las variables explican un 33.1% de la variacion
de ```G1```
- Solo el intercepto, ```studytime```, ```failures```, ```schoolsup```, ```famsup```, ```goout``` y ```health``` afectan significativamente a la variable ```G1```
- Obtuvo un Log-Likelihood de -683 

# Modelo para G2

In [9]:
modelo_g2 = aux3.crear_modelo_lineal(df.drop(columns=['G1', 'G3']), 'G2')
modelo_g2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     G2   R-squared:                       0.301
Model:                            OLS   Adj. R-squared:                  0.189
Method:                 Least Squares   F-statistic:                     2.693
Date:                Sat, 04 Feb 2023   Prob (F-statistic):           2.14e-06
Time:                        23:14:57   Log-Likelihood:                -729.98
No. Observations:                 284   AIC:                             1540.
Df Residuals:                     244   BIC:                             1686.
Df Model:                          39                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept         12.9915      4.063      3.197      0.002       4.988      20.995
school             0.2397      0.785      0.305      0.760      -1.306       1.785
sex                0.8885      0.508      1.748      0.082      -0.112       1.889
age               -0.0696      0.218     -0.320      0.749      -0.498       0.359
address           -0.7540      0.573     -1.315      0.190      -1.883       0.375
famsize            0.4639      0.518      0.895      0.372      -0.557       1.485
Pstatus            0.5952      0.766      0.777      0.438      -0.913       2.103
Medu               0.3221      0.332      0.971      0.332      -0.331       0.975
Fedu              -0.1232      0.278     -0.443      0.658      -0.671       0.425
traveltime        -0.2211      0.342     -0.646      0.519      -0.896       0.453
studytime          0.4171      0.289      1.444      0.150      -0.152       0.986
failures          -1.1293      0.333     -3.396      0.001      -1.784      -0.474
schoolsup         -1.0396      0.664     -1.566      0.119      -2.347       0.268
famsup             1.0023      0.484      2.071      0.039       0.049       1.956
paid               0.4781      0.474      1.008      0.315      -0.457       1.413
activities         0.2109      0.436      0.484      0.629      -0.647       1.069
nursery           -0.1439      0.536     -0.269      0.789      -1.199       0.911
higher            -1.1094      1.031     -1.076      0.283      -3.140       0.921
internet          -0.4415      0.590     -0.748      0.455      -1.605       0.722
romantic          -0.7324      0.470     -1.559      0.120      -1.658       0.193
famrel             0.0282      0.251      0.112      0.911      -0.466       0.522
freetime           0.3319      0.227      1.462      0.145      -0.115       0.779
goout             -0.9655      0.234     -4.130      0.000      -1.426      -0.505
Dalc              -0.2099      0.337     -0.624      0.533      -0.873       0.453
Walc               0.3593      0.247      1.455      0.147      -0.127       0.846
health            -0.3901      0.167     -2.342      0.020      -0.718      -0.062
absences          -0.0079      0.029     -0.274      0.785      -0.064       0.049
health_Mjob        1.3034      1.081      1.205      0.229      -0.827       3.433
other             -0.0448      0.704     -0.064      0.949      -1.432       1.342
services           0.5276      0.795      0.663      0.508      -1.039       2.094
teacher           -0.9595      0.984     -0.975      0.330      -2.898       0.979
health_Fjob       -0.3798      1.536     -0.247      0.805      -3.405       2.645
other_Fjob        -0.6572      1.159     -0.567      0.571      -2.940       1.625
services_Fjob      0.2459      1.165      0.211      0.833      -2.049       2.541
teacher_Fjob

Al realizar un modelo de regresion lineal con todas las variables y G2 como vector objetivo se
obtuvo lo siguiente:

- El modelo presenta un $R^2$ = 0.301. Por lo tanto, las variables explican un 30.1% de la variacion
de ```G2```
- Solo el intercepto, ```failures```, ```famsup```, ```goout``` y ```health``` afectan significativamente a la variable ```G2```
- Obtuvo un Log-Likelihood de -730

Por lo tanto, al comparar este modelo con el que predice G1, el modelo para predecir G1 es mejor.

# Modelo para G3

In [10]:
modelo_g3 = aux3.crear_modelo_lineal(df.drop(columns=['G1', 'G2']), 'G3')
modelo_g3.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     G3   R-squared:                       0.312
Model:                            OLS   Adj. R-squared:                  0.201
Method:                 Least Squares   F-statistic:                     2.831
Date:                Sat, 04 Feb 2023   Prob (F-statistic):           6.14e-07
Time:                        23:21:29   Log-Likelihood:                -780.76
No. Observations:                 284   AIC:                             1642.
Df Residuals:                     244   BIC:                             1787.
Df Model:                          39                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept         12.3799      4.859      2.548      0.011       2.809      21.951
school             0.9206      0.938      0.981      0.327      -0.927       2.769
sex                1.0108      0.608      1.663      0.098      -0.186       2.208
age               -0.1608      0.260     -0.618      0.537      -0.673       0.352
address           -0.8873      0.686     -1.294      0.197      -2.238       0.463
famsize            0.3597      0.620      0.580      0.562      -0.861       1.580
Pstatus            1.0103      0.916      1.103      0.271      -0.793       2.814
Medu               0.5017      0.397      1.265      0.207      -0.279       1.283
Fedu              -0.3060      0.333     -0.920      0.358      -0.961       0.349
traveltime        -0.1333      0.409     -0.326      0.745      -0.940       0.673
studytime          0.3779      0.345      1.094      0.275      -0.302       1.058
failures          -1.4767      0.398     -3.713      0.000      -2.260      -0.693
schoolsup         -1.0615      0.794     -1.337      0.182      -2.625       0.502
famsup             0.9381      0.579      1.621      0.106      -0.202       2.078
paid               0.5265      0.567      0.928      0.354      -0.591       1.644
activities         0.6244      0.521      1.199      0.232      -0.402       1.650
nursery           -0.0846      0.641     -0.132      0.895      -1.347       1.177
higher            -1.7295      1.233     -1.403      0.162      -4.158       0.699
internet          -0.4170      0.706     -0.591      0.555      -1.808       0.974
romantic          -0.8878      0.562     -1.581      0.115      -1.994       0.219
famrel             0.4857      0.300      1.621      0.106      -0.105       1.076
freetime           0.4721      0.271      1.739      0.083      -0.063       1.007
goout             -1.2396      0.280     -4.434      0.000      -1.790      -0.689
Dalc              -0.5493      0.402     -1.365      0.174      -1.342       0.243
Walc               0.7566      0.295      2.562      0.011       0.175       1.338
health            -0.3936      0.199     -1.976      0.049      -0.786      -0.001
absences           0.0336      0.034      0.977      0.329      -0.034       0.101
health_Mjob        1.3603      1.293      1.052      0.294      -1.187       3.907
other              0.2651      0.842      0.315      0.753      -1.393       1.924
services           0.7422      0.951      0.780      0.436      -1.131       2.616
teacher           -1.0447      1.177     -0.888      0.376      -3.362       1.273
health_Fjob       -0.6401      1.836     -0.349      0.728      -4.257       2.977
other_Fjob        -1.4023      1.386     -1.012      0.313      -4.132       1.327
services_Fjob     -0.5543      1.393     -0.398      0.691      -3.298       2.190
teacher_Fjob

Al realizar un modelo de regresion lineal con todas las variables y G3 como vector objetivo se
obtuvo lo siguiente:

- El modelo presenta un $R^2$ = 0.312. Por lo tanto, las variables explican un 31.2% de la variacion
de ```G3```
- Solo el intercepto, ```failures```, ```goout```, ```Walc``` y ```health``` afectan significativamente a la variable ```G3```
- Obtuvo un Log-Likelihood de -780

Por lo tanto, al comparar este modelo con los previos se obtienen las siguientes conclusiones:

- El modelo para predecir G3 es el que obtuvo el peor valor de Log-Likelihood (-780)
- El modelo para predecir G3 obtuvo un poder explicativo ($R^2$) mejor que el modelo para predecir G2, pero peor para predecir G1
- Solamente se utilizaran las variables ```failures```, ```goout```, ```Walc``` y ```health``` para el modelamiento predictivo